Library/Homebrew/utils/pypi.rb

```diff
@@ -0,0 +1,99 @@
+  end
+
+  def unable_to_resolve_deps_error(formula_name, odie_on_error)
+    return unless odie_on_error
+
+    odie "Unable to resolve some dependencies. Please update the resources for \"#{formula_name}\" manually."
+  end
+
+  def update_python_resources!(formula, print_only: false, silent: false, odie_on_error: true)
+    # PyPI package name isn't always the same as the formula name. Try to infer from the URL.
+    pypi_name = if formula.stable.url.start_with?("https://files.pythonhosted.org/packages/")
+      File.basename(formula.stable.url).match(/^(.+)-[a-z0-9.]+$/)[1]
+    else
+      formula.name
+    end
+
+    # Ensure there are no non-pypi resources
+    non_pypi_resources = formula.resources.reject do |resource|
+      resource.url.start_with? "https://files.pythonhosted.org/packages/"
+    end
+    return unable_to_resolve_deps_error formula.name, odie_on_error unless non_pypi_resources.empty?
+
+    # Get list of required packages
+    ohai "Retrieving PyPI dependencies for \"#{pypi_name}==#{formula.version}\"" unless print_only || silent
+    pipgrip_output = Utils.popen_read Formula["pipgrip"].bin/"pipgrip", "--json", "--no-cache-dir",
+                                      "#{pypi_name}==#{formula.version}"
+    return unable_to_resolve_deps_error formula.name, odie_on_error unless $CHILD_STATUS.success?
+
+    packages = JSON.parse(pipgrip_output).sort.to_h
+
+    # Remove extra packages that may be included in pipgrip output
+    exclude_list = %w[argparse pip setuptools wsgiref] + [pypi_name]
+    packages.delete_if do |p|
+      exclude_list.include? p
+    end
+
+    # Create resource blocks
+    new_resource_blocks = ""
+    packages.each do |package, version|
+      name, url, checksum = get_pypi_info package, version
+      return unable_to_resolve_deps_error formula.name, odie_on_error if name.nil?
+
+      # Indent and append each resource block
+      new_resource_blocks += "  " + <<~EOS.chomp.gsub("\n", "\n  ") + "\n\n"
+        resource "#{name}" do
+          url "#{url}"
+          sha256 "#{checksum}"
+        end
+      EOS
+    end
+
+    # Print resource lines only if --print-only is passed
+    puts new_resource_blocks.chomp if print_only
+    return if print_only
+
+    # Check whether resources already exist (excluding homebrew-virtualenv)
+    if formula.resources.empty? ||
+       (formula.resources.length == 1 && formula.resources.first.name == "homebrew-virtualenv")
+      # Place resources above install method
+      inreplace_regex = /  def install/
+      new_resource_blocks += "  def install"
+    else
+      # Replace existing resource blocks with new resource blocks
+      inreplace_regex = /  (resource .* do\s+url .*\s+sha256 .*\s+ end\s*)+/
+      new_resource_blocks += "  "
+    end
+
+    ohai "Updating resource blocks" unless silent
+    Utils::Inreplace.inreplace formula.path do |s|
+      s.sub! inreplace_regex, new_resource_blocks
+    end
+
+    true
```

Comment:
This does not seem to be used anywhere if I see correctly.

Reply:
This is leftover from some previous attempts at dealing with errors. I'm going to try to rethink the logic here, so this will either become used or go away.